pallets/liquidation/src/lib.rs

```diff
@@ -0,0 +1,331 @@
+		///
+		/// Emits `Liquidated` event when successful.
+		///
+		#[pallet::call_index(0)]
+		#[pallet::weight(<T as Config>::WeightInfo::liquidate()
+			.saturating_add(<T as Config>::RouterWeightInfo::sell_weight(route))
+			.saturating_add(<T as Config>::RouterWeightInfo::get_route_weight())
+			.saturating_add(<T as Config>::GasWeightMapping::gas_to_weight(<T as Config>::GasLimit::get(), true))
+		)]
+		pub fn liquidate(
+			origin: OriginFor<T>,
+			collateral_asset: AssetId,
+			debt_asset: AssetId,
+			user: EvmAddress,
+			debt_to_cover: Balance,
+			route: Vec<Trade<AssetId>>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			let pallet_acc = Self::account_id();
+
+			// `route` needs to be provided in order to calculate correct TX fee.
+			// Ensure that the provided route matches the route we get from the router.
+			ensure!(
+				route
+					== T::Router::get_route(AssetPair {
+						asset_in: collateral_asset,
+						asset_out: debt_asset,
+					}),
+				Error::<T>::InvalidRoute
+			);
+
+			let debt_asset_initial_balance = <T as Config>::Currency::balance(debt_asset, &pallet_acc);
+			let collateral_asset_initial_balance = <T as Config>::Currency::balance(collateral_asset, &pallet_acc);
+
+			// mint borrow asset
+			<T as Config>::Currency::mint_into(debt_asset, &pallet_acc, debt_to_cover)?;
+
+			// liquidation
+			let pallet_evm_account = T::EvmAccounts::evm_address(&pallet_acc);
+			let mm_contract_address = T::MoneyMarketContract::get();
+
+			let context = CallContext::new_call(mm_contract_address, pallet_evm_account);
+			let collateral_asset_evm_address =
+				T::Erc20Mapping::encode_evm_address(collateral_asset).ok_or(Error::<T>::AssetConversionFailed)?;
+			let debt_asset_evm_address =
+				T::Erc20Mapping::encode_evm_address(debt_asset).ok_or(Error::<T>::AssetConversionFailed)?;
+			let data = Self::encode_liquidation_call_data(
+				collateral_asset_evm_address,
+				debt_asset_evm_address,
+				user,
+				debt_to_cover,
+				false,
+			);
+
+			// EVM call
+			let value = U256::zero();
+			let gas = T::GasLimit::get();
+			let (exit_reason, value) = T::Evm::call(context, data, value, gas);
+			if exit_reason != ExitReason::Succeed(ExitSucceed::Returned) {
+				log::debug!(target: "liquidation",
+					"Evm execution failed. Reason: {:?}", value);
+				return Err(Error::<T>::EvmExecutionFailed.into());
+			}
+
+			// swap collateral asset for borrow asset
+			let collateral_earned = <T as Config>::Currency::balance(collateral_asset, &pallet_acc)
+				.checked_sub(collateral_asset_initial_balance)
+				.ok_or(ArithmeticError::Overflow)?;
+			T::Router::sell(
+				RawOrigin::Signed(pallet_acc.clone()).into(),
+				collateral_asset,
+				debt_asset,
+				collateral_earned,
+				1,
+				route.clone(),
+			)?;
+
+			//burn
+			let debt_asset_final_balance = <T as Config>::Currency::balance(debt_asset, &pallet_acc);
+			let debt_asset_earned = debt_asset_final_balance
```

Comment:
Maybe fire normal error here??

Reply:
aggree, if there is no profit it would underflow, we should have special error for that